great_expectations/experimental/datasources/sources.py

```diff
@@ -199,12 +211,84 @@ def _add_asset_factory(
             _add_asset_factory.__signature__ = _merge_signatures(  # type: ignore[attr-defined]
                 _add_asset_factory, asset_type, exclude={"type"}
             )
-            setattr(ds_type, asset_factory_method_name, _add_asset_factory)
+            setattr(ds_type, add_asset_factory_method_name, _add_asset_factory)
+
+            def _read_asset_factory(self: Datasource, **kwargs) -> pydantic.BaseModel:
+                existing_asset_names: Generator[str, None, None] = (
+                    asset_name for asset_name in self.assets.keys()
+                )
+                asset_name_prefix = f"{asset_type_name}_asset_"
+                max_asset_number = 0
+                for asset_name in existing_asset_names:
+                    try:
+                        # this can fail for an asset named like csv_asset_foo
+                        asset_number = int(asset_name.split("_")[-1])
+                    except ValueError:
+                        asset_number = 1
+                    if (
+                        asset_name.startswith(asset_name_prefix)
+                        and asset_number > max_asset_number
+                    ):
+                        max_asset_number = asset_number
+
+                name = asset_name_prefix + str(max_asset_number + 1)
+                asset = asset_type(name=name, **kwargs)
+                return self.add_asset(asset)
+
+            _read_asset_factory.__signature__ = _merge_signatures(  # type: ignore[attr-defined]
+                _read_asset_factory, asset_type, exclude={"type"}
+            )
+            read_asset_factory_method_name = f"read_{asset_type_name}"
+            setattr(ds_type, read_asset_factory_method_name, _read_asset_factory)
+
         else:
             logger.debug(
-                f"`{asset_factory_method_name}()` already defined `{ds_type.__name__}`"
+                f"`{add_asset_factory_method_name}()` already defined `{ds_type.__name__}`"
+            )
+
+    @property
+    def pandas_default(self) -> PandasDatasource:
```

Comment:
@NathanFarmer @Kilo59 While I know how to work with `_PandasDatasource` in order to give it support for cloud/network-based file storage environments, I would like to get your ideas for how we can do the same for this `PandasDatasource` (the default one).  Thank you.

Reply:
I will need to learn more about what the data connector supports and what the requirements are, because we are supporting all pandas reader methods (except `read_fwf`), which already have some level of support for the cloud environments.

Reply:
@NathanFarmer The questions we will need to ponder are more related to our architecture (the `DataConnector` does not interfere with the `DataFrame` or any other `ExecutionEngine` behaviors).  Thanks.